### ISSUE TITLE
Set-PASUser - Unix time error on ExpiryDate

### ISSUE THREAD

**Issue by anycard94** [pspete/psPAS#536]
Set-PASUser in psPAS 6.4.80 using CyberArk PAM 12.6 on accounts without expiryDate shows the following error:
```
"ConvertTo-UnixTime : The input object cannot be bound to any parameters for the command either because the command does not take pipeline input
or the input and its properties do not match any of the parameters that take pipeline input.
At line:44 char:65
+ ... ct['ExpiryDate'] = $UserProperties['ExpiryDate'] | ConvertTo-UnixTime
+                                                        ~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (-62135578800:Int64) [ConvertTo-UnixTime], ParameterBindingException
    + FullyQualifiedErrorId : InputObjectNotBound,ConvertTo-UnixTime"
```
This error doesn't prevent the PAS user changes but is thrown for parameter changed using Set-PASUser.

Environment:

* PowerShell Version: 5.1.22621.2506
* psPAS Version: 6.4.80
* CyberArk Version: 12.6.12.217

The accounts in question all have expiryDate of -62135578800 which is a nullable date. If we set an account with a valid expiration date it does not throw this error.

I was able to modify Format-PASUserObject.ps1 to do a check of negative numbers to remove this error:
```
'ExpiryDate' {
				#Include date string in required format
				if ($UserProperties['ExpiryDate'] -le -1){
					$UserObject['ExpiryDate'] = $null
				} else {
					$UserObject['ExpiryDate'] = UserProperties['ExpiryDate'] | ConvertTo-UnixTime 
				}
	
			}
```
This removes the error and updates the account. However not sure if this is a proper fix but should be a good starting point. 

I tried other fixes such as changing from piping to ConvertTo-UnixTime to specifying ConvertTo-UnixTime -Date UserProperties['ExpiryDate'] but that throws an error about DateTime.MinValue.Ticks / Max.Ticks and doesn't update PAS user.

**Comment by pspete** [comment]
Hi @anycard94 - thanks for the report.
This is the first time hearing of the value `-62135578800` in relation to the expiry date for a user account in CyberArk.
To remove an expiry date one would usually provide a value of UnixTime of 0 (or a datetime of 1/1/1970).

**Comment by pspete** [comment]
its not something we see when creating a user nor are able to replicate
![image](https://github.com/pspete/psPAS/assets/10600090/8da826f2-85aa-4c20-bf47-6b2320d6f9bc)
Expiry date is not set on creation when no value is specified.

`ExpiryDate` parameter of `New-PASUser` accepts a DateTime object type, and sets the expected value when specified
![image](https://github.com/pspete/psPAS/assets/10600090/97e65988-6aab-41cb-a2be-a7408b171dbe)

Providing a DateTime of 1/1/1970 clears the expirydate when specified for `Set-PASUser`
![image](https://github.com/pspete/psPAS/assets/10600090/7bcddd30-ce2f-48ba-ae5b-7f49474efa5c)
psPAS converts the date to unix time and provides a value of 0 in the request

**Comment by anycard94** [comment]
I must have something different going on with date time on my end. If I set expiry date to 1/1/1970 it returns value 18000 instead of 0.

Creating user with expire date of tomorrow
![image](https://github.com/pspete/psPAS/assets/169183501/f6f02940-9674-4636-9888-c658ccec7ed3)

Changing expiry date to 1/1/1970
![image](https://github.com/pspete/psPAS/assets/169183501/d6ef72cb-a790-4109-b480-c7d534990c59)
![image](https://github.com/pspete/psPAS/assets/169183501/852180bd-488d-445b-a731-50b828622fdc)

**Comment by anycard94** [comment]
The time I'm getting is UTC - I wrote a function long ago to convert:

![image](https://github.com/pspete/psPAS/assets/169183501/983c2693-df03-456f-bd43-c0e4c9e58bd2)

**Comment by pspete** [comment]
Maybe something to do with timezone offset then - presume you are UTC-5?

**Comment by pspete** [comment]
Potentially due to the `ToUniversalTime()` conversion here....
https://github.com/pspete/psPAS/blob/b8fba0fbf3c1acab2e958f275f39a73647a4b129/psPAS/Private/ConvertTo-UnixTime.ps1#L37

but - removing this would impact times used/returned by other functions in the module.

does `(Get-Date 1/1/1970).AddHours(-5)` land you at 0 (when updating expirydate)?

**Comment by anycard94** [comment]
@pspete - I tried some modifications but probably not in the preferred way. It appears to give the correct results. Please feel free to copy / modify as needed. 


Created a new file ConvertTo-UtcTime.ps1
```
Function ConvertTo-UtcTime {
	[CmdletBinding()]
	[OutputType('System.Integer')]
	Param(
		[Parameter(
			Mandatory = $true,
			ValueFromPipeline = $true
		)]
		[int64]$time
	)
	begin {
		$currentCulture = [System.Threading.Thread]::CurrentThread.CurrentCulture
	}
	process {
		[System.Threading.Thread]::CurrentThread.CurrentCulture = 'en-US'
		
		#Convert EPOCH Time from local
		$UTCTime = (Get-Date 01/01/1970)+([System.TimeSpan]::fromseconds($time))
		$strCurrentTimeZone = (Get-WmiObject win32_timezone).StandardName
		$TZ = [System.TimeZoneInfo]::FindSystemTimeZoneById($strCurrentTimeZone)
		$UTCTimeFinal = [System.TimeZoneInfo]::ConvertTimeFromUtc($UTCTime, $TZ)
		#If the time is before 1/1/1970 set final UTC to 1/1/1970
		if($UTCTimeFinal -le (Get-Date 1/1/1970)){
			$UTCTimeFinal = (Get-Date 1/1/1970)
		}
		
		# Do UTC Offset
		$UtcOffSetHours = $TZ.BaseUtcOffset.Hours
		$UtcOffSetMinutes = $TZ.BaseUtcOffset.Minutes
                # Are we currently in Daylight Savings??
		If($tz.IsDaylightSavingTime((get-date))){$UtcOffSetHours += 1}
		
		# Return Value
		Return $UTCTimeFinal.AddHours($UtcOffSetHours).AddMinutes($UtcOffSetMinutes)
	}
	end {
		[System.Threading.Thread]::CurrentThread.CurrentCulture = $currentCulture
	}
}
```

Modified Format-PASUserObject.ps1 - I added the switch as my expire date of -62135578800 comes back as Int64 type instead of DateTime.

```
switch ($UserProperties.keys) {

			'ExpiryDate' {
				#Include date string in required format
				switch ($UserProperties['ExpiryDate'].GetType().Name){
					'DateTime'{$UserObject['ExpiryDate'] = $UserProperties['ExpiryDate'] | ConvertTo-UnixTime}
					'Int64' {$UserObject['ExpiryDate'] = ($UserProperties['ExpiryDate'] | ConvertTo-UtcTime) | ConvertTo-UnixTime}
				}
			}

```

Modified ConvertTo-UnixTime.ps1 to modify date on offset.
```
process {
		[System.Threading.Thread]::CurrentThread.CurrentCulture = 'en-US'
		
		$strCurrentTimeZone = (Get-WmiObject win32_timezone).StandardName
		$TZ = [System.TimeZoneInfo]::FindSystemTimeZoneById($strCurrentTimeZone)
		# Do UTC Offset
		$UtcOffSetHours = $TZ.BaseUtcOffset.Hours
		$UtcOffSetMinutes = $TZ.BaseUtcOffset.Minutes
		# Are we currently in Daylight Savings??
		If($tz.IsDaylightSavingTime((get-date))){$UtcOffSetHours += 1}

		# Return Value
		$Date = $Date.AddHours($UtcOffSetHours).AddMinutes($UtcOffSetMinutes)
		
		
		$UnixTime = [math]::Round($(Get-Date $Date.ToUniversalTime() -UFormat %s))

		If ($Milliseconds) {
			$UnixTime = $UnixTime * 1000
		}

		$UnixTime

```

Results

Before ConvertTo-UnixTime.ps1 fix for TZ

```
PS C:\> date

Tuesday, May 14, 2024 2:45:24 PM


PS C:\> Set-PASUser -id 144 -username cybtest -ExpiryDate (get-date).addhours(4)

ID  UserName Source   UserType Suspended enableUser ExpiryDate            Location
--  -------- ------   -------- --------- ---------- ----------            --------
144 CybTest  CyberArk EPVUser  False     True       5/14/2024 10:45:29 PM \


PS C:\> date

Tuesday, May 14, 2024 2:46:17 PM
```

After modifications ConvertTo-UnixTime.ps1 for TZ

```
PS C:\> date

Tuesday, May 14, 2024 3:10:08 PM


PS C:\> Set-PASUser -id 144 -username cybtest -description 'test' -ExpiryDate (Get-Date).AddHours(4)

ID  UserName Source   UserType Suspended enableUser ExpiryDate           Location
--  -------- ------   -------- --------- ---------- ----------           --------
144 CybTest  CyberArk EPVUser  False     True       5/14/2024 7:10:11 PM \


PS C:\> date

Tuesday, May 14, 2024 3:10:14 PM

```

**Comment by pspete** [comment]
Nice - will take a look next time I have some dev time 👍

**Comment by pspete** [comment]
Should now be resolved in 6.4.85.
Set-PASUser now does not attempt to convert in-situ property values.
Unix epoch time is also not converted to universal time if specified as an expirydate.
